In [1]:
import numpy as np
import pandas as pd
import pickle
from sklearn.preprocessing import MinMaxScaler 
import math
from itertools import combinations
from scipy.spatial import ConvexHull

In [2]:
raw_data = pd.read_csv('Clubs_D50.csv')
raw_data.head()

District Division  Area       Club                Clubname    Clubest  \
0      50.0        E  11.0     5509.0        Twin Cities Club   6/1/1984   
1      50.0        E  11.0     7036.0         University Club   6/1/1988   
2      50.0        E  12.0   584009.0         Conquerors Club  7/11/2003   
3      50.0        E  11.0     9682.0  200 Watt Speakers Club  11/1/1993   
4      50.0        E  11.0  1100434.0              Solid Rock   4/1/2008   

  Clubstatus  DCP  Members       Clubcity  ...        Clubphone     Clubday_o  \
0     Active  NaN      NaN      Texarkana  ...  +1 903-278-6043      Thursday   
1     Active  NaN      NaN      Texarkana  ...  +1 903-244-5720       Mondays   
2     Active  NaN      NaN  Barksdale AFB  ...  +1 318-294-1038       Mondays   
3     Active  NaN      NaN     Shreveport  ...  +1 318-349-0767  3rd Saturday   
4     Active  NaN      NaN      Texarkana  ...  +1 903-793-2015       Tuesday   

        Clubtime_o Clubday  Clubtime Clublen Clubfreq Clubopen  Clubadvanced  \
0         12:00 PM     Thu  12:00:00     0.0       W1        O           NaN   
1          6:30 PM     Mon  18:30:00     0.0       W1        O           NaN   
2       12:15 p.m.     Mon  12:15:00     0.0       W1        O           Yes   
3          1:00 PM     Sat  13:00:00     0.0       M1        O           Yes   
4  6:30 pm-7:30 pm     Tue  18:30:00    60.0       W1        O           NaN   

  Notes:  
0    NaN  
1    NaN  
2    NaN  
3    NaN  
4    NaN  

[5 rows x 23 columns]

In [15]:
math.ceil(len(raw_data) / 5)
(len(raw_data) / 5)

57.4

In [146]:
sorted_data = raw_data.sort_values(by = "Club")

In [147]:
sorted_data.head()

District Division  Area    Club                            Clubname  \
174      50.0        S  53.0   437.0               Literary Legends Club   
81       50.0        M  31.0   713.0                Big "D" Toastmasters   
6        50.0        E  12.0   718.0                     Shreveport Club   
226      50.0        T  65.0   967.0  Ericsson Village Toastmasters Club   
187      50.0        S  55.0  1165.0      City Talkers Toastmasters Club   

       Clubest  Clubstatus  DCP  Members    Clubcity  ...        Clubphone  \
174   5/1/2002      Active  NaN      NaN      Dallas  ...  +1 214-467-6798   
81    3/1/1949      Active  NaN      NaN      Dallas  ...  +1 214-660-5483   
6     5/1/1949      Active  NaN      NaN  Shreveport  ...  +1 318-841-3129   
226  8/31/2001      Active  NaN      NaN       Plano  ...  +1 972-583-3389   
187   4/1/1998  Ineligible  NaN      NaN      Dallas  ...  +1 214-671-8875   

           Clubday_o  Clubtime_o Clubday  Clubtime Clublen Clubfreq Clubopen  \
174         Saturday     9:15 AM     Sat   9:15:00     0.0       W1        O   
81          Thursday     6:30 PM     Thu  18:30:00     0.0       W1        O   
6             Monday    12:00 PM     Mon  12:00:00     0.0       W1        O   
226         Thursday    12:00 PM     Thu  12:00:00     0.0       W1        O   
187  Tues, 1st & 3rd  12:00 p.m.     Tue  12:00:00     0.0       M2        O   

     Clubadvanced Notes:  
174           NaN    NaN  
81            NaN    NaN  
6             NaN    NaN  
226           NaN    NaN  
187           NaN    NaN  

[5 rows x 23 columns]

In [23]:
with open('../data/bayg29-loc.pickle', 'rb') as f:  
    loc = pickle.load(f)

In [24]:
loc[:5]

[array([1150., 1760.], dtype=float32),
 array([ 630., 1660.], dtype=float32),
 array([  40., 2090.], dtype=float32),
 array([ 750., 1100.], dtype=float32),
 array([ 750., 2030.], dtype=float32)]

In [25]:
with open('../data/bayg29-dist.pickle', 'rb') as f:  
    dist = pickle.load(f)

In [46]:
%timeit dist[0][1]

107 ns ± 38.3 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [30]:
dist_array = np.array(dist)

In [42]:
dist_array[0]

array([   0.        ,  529.52807617, 1158.01550293,  771.75128174,
        482.59713745,  332.41540527, 1217.4152832 ,  364.00549316,
        616.11688232,  629.36474609, 1249.07971191,  540.37023926,
        456.94638062, 1238.2244873 ,  948.47247314,  574.89129639,
       1488.38842773, 1134.06347656,  817.43499756,  671.19299316,
        320.15621948, 1422.39233398,  864.00231934,  282.31188965,
        978.67254639,  756.63726807,  460.1086731 ,  186.01075745,
        820.06097412])

In [60]:
n_dist = MinMaxScaler().fit_transform(dist)

In [61]:
n_dist[0]

array([0.        , 0.36889492, 0.53611253, 0.60701992, 0.29291919,
       0.19758648, 0.56361213, 0.22280112, 0.33754381, 0.54795777,
       0.70139818, 0.27692366, 0.38246559, 0.71541408, 0.64890846,
       0.37664807, 0.76275475, 0.706352  , 0.53904754, 0.51919822,
       0.23064479, 0.73922687, 0.43404052, 0.20540603, 0.5447504 ,
       0.40237555, 0.29303906, 0.11109724, 0.44259779])

In [47]:
%timeit dist_array[0][1]

484 ns ± 50.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [38]:
%timeit combinations(['A', 'B', 'C', 'D', 'E'], 2)

373 ns ± 29.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [41]:
len(list(combinations(['A', 'B', 'C', 'D', 'E'], 2)))

10

In [54]:
dcp = pd.read_csv('20200630_D50_Performance.csv')

In [57]:
dcp['Club'].dtype

dtype('float64')

In [58]:
awards = pd.read_csv('20201229_D50_Awards.csv')

In [59]:
awards['Club'].dtype

dtype('int64')

In [16]:
club_zips = pd.read_csv('club_zips.csv')

In [17]:
club_zips.head()

club_no         zip      lat     long
0     5509       75501  33.3672 -94.2356
1     7036       75501  33.3672 -94.2356
2   584009       71110  32.5153 -93.6345
3     9682       71101  32.5072 -93.7444
4  1100434  75501-6314  33.3672 -94.2356

The pattern I need to match: `self.locations.append(np.asarray(row, dtype=np.float32)`

An example: `array([1150., 1760.], dtype=float32)`. Lon = x, Lat = y

In [21]:
np.asarray(club_zips.iloc[0], dtype=np.float32)

array([ 5.50900e+03,  7.55010e+04,  3.33672e+01, -9.42356e+01],
      dtype=float32)

In [35]:
[np.asarray([row[4], row[3]], dtype=np.float32) for row in club_zips.head().itertuples()]

[array([-94.2356,  33.3672], dtype=float32),
 array([-94.2356,  33.3672], dtype=float32),
 array([-93.6345,  32.5153], dtype=float32),
 array([-93.7444,  32.5072], dtype=float32),
 array([-94.2356,  33.3672], dtype=float32)]

In [34]:
for row in club_zips.head().itertuples():
    print(row[4], row[3])

-94.2356 33.3672
-94.2356 33.3672
-93.6345 32.5153
-93.7444 32.5072
-94.2356 33.3672


In [74]:
with open('../data/loc.pickle', 'rb') as f:  
    locations = pickle.load(f)

In [77]:
locations[:5]

[array([-96.8742,  32.6623], dtype=float32),
 array([-96.7717,  32.8335], dtype=float32),
 array([-93.7444,  32.5072], dtype=float32),
 array([-96.8128,  33.0838], dtype=float32),
 array([-96.7999,  32.7877], dtype=float32)]

In [86]:
len(locations) > 0

True

In [89]:
with open('../data/dist.pickle', 'rb') as f:  
    distances = pickle.load(f)

In [90]:
distances[0]

array([0.      , 0.05865 , 0.979   , 0.1225  , 0.04257 , 0.075   ,
       0.10956 , 0.02286 , 0.0916  , 0.0986  , 0.01895 , 0.0736  ,
       0.7446  , 0.0976  , 0.08276 , 0.05865 , 0.086   , 0.0378  ,
       0.04013 , 0.04257 , 0.0986  , 0.11255 , 0.11255 , 0.1225  ,
       0.075   , 0.075   , 0.8486  , 0.1649  , 0.0986  , 0.086   ,
       0.1284  , 0.974   , 0.0406  , 0.1515  , 0.98    , 0.0986  ,
       0.086   , 0.955   , 0.0976  , 0.04257 , 0.04013 , 0.09644 ,
       0.1225  , 0.08765 , 0.1179  , 0.11346 , 0.07434 , 0.04013 ,
       0.1225  , 0.98    , 0.1225  , 0.1372  , 0.04257 , 0.11255 ,
       0.0854  , 0.1225  , 0.8486  , 0.827   , 0.04507 , 0.1225  ,
       0.08185 , 0.1344  , 0.0854  , 0.05865 , 0.0986  , 0.05518 ,
       0.0406  , 0.04013 , 0.0779  , 0.11255 , 0.04257 , 0.0986  ,
       0.0986  , 0.3164  , 0.372   , 0.1936  , 0.979   , 0.9805  ,
       0.086   , 0.1038  , 0.98    , 0.04013 , 0.1284  , 0.2942  ,
       0.1093  , 0.11255 , 0.1284  , 0.1093  , 0.1225  , 0.153

In [56]:
a_dist = np.linalg.norm(locations[0] - locations[1])
float16_dist = a_dist.astype('float16')
print(f'Base: {a_dist}')
print(f'Floa: {float16_dist}')

Base: 0.19953899085521698
Floa: 0.1995849609375


In [71]:
size = len(locations)
distances = np.zeros((size, size), dtype='float16')
np.zeros((size, size), dtype=np.float16)

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float16)

In [60]:
distances.shape

(221, 221)

In [61]:
# Populate the matrix
for i in range(size):
    for j in range(i + 1, size):
        # Get the distance between them in degrees
        distance = np.linalg.norm(locations[j] - locations[i])
        distances[i][j] = distance
        distances[j][i] = distance

In [67]:
distances[0][:10]

array([0.    , 0.1996, 3.133 , 0.426 , 0.1458, 0.2534, 0.3525, 0.0806,
       0.316 , 0.334 ], dtype=float16)

In [88]:
MinMaxScaler().fit_transform(distances)

array([[0.     , 0.05865, 0.979  , ..., 0.1372 , 0.978  , 0.1225 ],
       [0.05716, 0.     , 0.9517 , ..., 0.0941 , 0.9507 , 0.07294],
       [0.897  , 0.895  , 0.     , ..., 0.9004 , 0.01114, 0.898  ],
       ...,
       [0.1398 , 0.0984 , 1.     , ..., 0.     , 0.9995 , 0.0271 ],
       [0.8887 , 0.886  , 0.01105, ..., 0.892  , 0.     , 0.8896 ],
       [0.122  , 0.0746 , 0.976  , ..., 0.0265 , 0.975  , 0.     ]],
      dtype=float16)

In [92]:
list(range(11))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [93]:
from scipy.spatial import ConvexHull, convex_hull_plot_2d

In [116]:
%timeit hull = ConvexHull(locations[:5])

1 ms ± 11.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [132]:
%timeit locations[:5]

154 ns ± 1.71 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [168]:
[locations[i] for i in [0, 1, 200, 3, 4]]

[array([-96.8742,  32.6623], dtype=float32),
 array([-96.7717,  32.8335], dtype=float32),
 array([-96.7999,  32.7877], dtype=float32),
 array([-96.8128,  33.0838], dtype=float32),
 array([-96.7999,  32.7877], dtype=float32)]

In [134]:
locations[:5]

[array([-96.8742,  32.6623], dtype=float32),
 array([-96.7717,  32.8335], dtype=float32),
 array([-93.7444,  32.5072], dtype=float32),
 array([-96.8128,  33.0838], dtype=float32),
 array([-96.7999,  32.7877], dtype=float32)]

In [94]:
hull

In [117]:
print(f'Input: {locations[:5]}')
print(f'Points: ')
print(hull.points)
print(f'Vertices: {hull.vertices}')
print(f'Area: {hull.area}')
print(f'Volume: {hull.volume}')

Input: [array([-96.8742,  32.6623], dtype=float32), array([-96.7717,  32.8335], dtype=float32), array([-93.7444,  32.5072], dtype=float32), array([-96.8128,  33.0838], dtype=float32), array([-96.7999,  32.7877], dtype=float32)]
Points: 
[[-96.87419891  32.66230011]
 [-96.771698    32.83349991]
 [-93.74440002  32.50719833]
 [-96.81279755  33.08380127]
 [-96.79989624  32.78770065]]
Vertices: [0 2 3]
Area: 6.681693771366869
Volume: 0.6643686612078376


In [105]:
locations[0][0]

-96.8742

In [160]:
def chunks(clubs, n):
    """Yield successive n-sized chunks from l."""
    for i in range(0, len(clubs), n):
        yield clubs[i:i + n]

test_areas_list = list(chunks(list(range(220)), 5))
test_areas_list[:5]

[[0, 1, 2, 3, 4],
 [5, 6, 7, 8, 9],
 [10, 11, 12, 13, 14],
 [15, 16, 17, 18, 19],
 [20, 21, 22, 23, 24]]

In [161]:
from math import pi

def compactness(area):
    hull = ConvexHull([locations[i] for i in area])
    return (4 * pi * hull.volume) / (hull.area * hull.area)

In [166]:
def average_compactness(areas):
    return np.mean([compactness(area) for area in areas])

In [167]:
average_compactness(test_areas_list)

0.37935183352638274

In [127]:
def get_area_perimeter(clubs):
        # Distance between last and first club
        distance = np.linalg.norm(locations[clubs[-1]] - locations[clubs[0]])
        #distance = distances[clubs[-1]][clubs[0]]
        # Add the distance between the rest
        for i in range(len(clubs) - 1):
            distance += np.linalg.norm(locations[clubs[i]] - locations[clubs[i+1]])
            #distance += distances[clubs[i]][clubs[i + 1]]
        return distance

In [129]:
%timeit get_area_perimeter([0, 1, 2, 3, 4])

44.2 µs ± 118 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [128]:
get_area_perimeter(list(range(5)))

6.808618

Conclude that get_area_perimeter is calculating the same thing as the perimeter of the hull. Skip that and go to the compactness based on the area and perimeter.

In [ ]:
# Keep just in case:
'''
    def get_area_perimeter(self, clubs):
        """
        Pseudo-calculates the length of the perimeter of an area (i.e. distance).
        If the points aren't in a circular order, it might zig-zag across. The
        GA will optimize the best order even though that's not important. The 
        performance loss of using geospatial modules to get the order right as 
        a proper polygon is not worth the savings in optimization. This is how
        route lengths would otherwise be calculated.
        :param clubs: A list of clubs that define the perimeter (route).
        """
        # Distance between last and first club
        distance = self.distances[clubs[-1]][clubs[0]]

        # Add the distance between the rest
        for i in range(len(clubs) - 1):
            distance += self.distances[clubs[i]][clubs[i + 1]]

        return distance
'''

In [148]:
np.mean([self.compactness(area) for area in areas_list])

District Division  Area    Club                            Clubname  \
174      50.0        S  53.0   437.0               Literary Legends Club   
81       50.0        M  31.0   713.0                Big "D" Toastmasters   
6        50.0        E  12.0   718.0                     Shreveport Club   
226      50.0        T  65.0   967.0  Ericsson Village Toastmasters Club   
187      50.0        S  55.0  1165.0      City Talkers Toastmasters Club   

       Clubest  Clubstatus  DCP  Members    Clubcity  ...        Clubphone  \
174   5/1/2002      Active  NaN      NaN      Dallas  ...  +1 214-467-6798   
81    3/1/1949      Active  NaN      NaN      Dallas  ...  +1 214-660-5483   
6     5/1/1949      Active  NaN      NaN  Shreveport  ...  +1 318-841-3129   
226  8/31/2001      Active  NaN      NaN       Plano  ...  +1 972-583-3389   
187   4/1/1998  Ineligible  NaN      NaN      Dallas  ...  +1 214-671-8875   

           Clubday_o  Clubtime_o Clubday  Clubtime Clublen Clubfreq Clubopen  \
174         Saturday     9:15 AM     Sat   9:15:00     0.0       W1        O   
81          Thursday     6:30 PM     Thu  18:30:00     0.0       W1        O   
6             Monday    12:00 PM     Mon  12:00:00     0.0       W1        O   
226         Thursday    12:00 PM     Thu  12:00:00     0.0       W1        O   
187  Tues, 1st & 3rd  12:00 p.m.     Tue  12:00:00     0.0       M2        O   

     Clubadvanced Notes:  
174           NaN    NaN  
81            NaN    NaN  
6             NaN    NaN  
226           NaN    NaN  
187           NaN    NaN  

[5 rows x 23 columns]